File: src/main/java/learning_flink_java/course07_time_windows/WindowsApp.java
package learning_flink_java.course07_time_windows;

import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.util.Collector;

public class WindowsApp {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        tumblingWindows(env);
        env.execute();
    }

    public static void tumblingWindows(StreamExecutionEnvironment env) {
        DataStreamSource<String> data = env.socketTextStream("localhost", 9999);
        data.flatMap(new FlatMapFunction<String, Tuple2<String, Integer>>() {
            @Override
            public void flatMap(String value, Collector<Tuple2<String, Integer>> out) throws Exception {
                String[] words = value.toLowerCase().split(" ");
                for (String word : words) {
                    out.collect(new Tuple2<>(word, 1));
                }
            }
        }).keyBy(0)
                .timeWindow(Time.seconds(5))
                .sum(1).print();
    }

    public static void slidingWindows(StreamExecutionEnvironment env) {
        DataStreamSource<String> data = env.socketTextStream("localhost", 9999);
        data.flatMap(new FlatMapFunction<String, Tuple2<String, Integer>>() {
            @Override
            public void flatMap(String value, Collector<Tuple2<String, Integer>> out) throws Exception {
                String[] words = value.toLowerCase().split(" ");
                for (String word : words) {
                    out.collect(new Tuple2<>(word, 1));
                }
            }
        }).keyBy(0)
                .timeWindow(Time.seconds(5), Time.seconds(10))
                .sum(1).print();
    }
}

File: src/main/scala/com/baich/learning_bigdata/learning_flink/course07_time_windows/WindowsApp.scala
package com.baich.learning_bigdata.learning_flink.course07_time_windows

import org.apache.flink.streaming.api.scala.{DataStream, StreamExecutionEnvironment}
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.windowing.assigners.{TumblingEventTimeWindows, TumblingProcessingTimeWindows}
import org.apache.flink.streaming.api.windowing.time.Time

/**
  * 滚动和滑动窗口调用的是timeWindow方法的异构
  * 调用时，slide参数传入时为滑动窗口，不传入时为滚动窗口
  */
object WindowsApp {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    //    tumblingWindows(env)
    slidingWindows(env)
    env.execute()
  }

  def tumblingWindows(env: StreamExecutionEnvironment): Unit = {
    val data: DataStream[String] = env.socketTextStream("localhost", 9999)
    data.flatMap(_.toLowerCase().split(" "))
      .map((_, 1))
      .keyBy(0)
      .timeWindow(Time.seconds(5))
      .sum(1)
      .print()
  }

  def slidingWindows(env: StreamExecutionEnvironment): Unit = {
    val data: DataStream[String] = env.socketTextStream("localhost", 9999)
    data.flatMap(_.toLowerCase().split(" "))
      .map((_, 1))
      .keyBy(0)
      .timeWindow(Time.seconds(10), Time.seconds(5))
      .sum(1)
      .print()
  }

}
